### ISSUE TITLE
NavigationAgent2D does not apply radius settings around NavigationPolygonInstance in 3.5 beta 4

### ISSUE THREAD

**Issue by nathanielhudson** [godotengine/godot#60546]
### Godot version

v3.5.beta4.official [b6968ab06]

### System information

Windows 10

### Issue description

Hello. Apologies if I've misunderstood anything and/or have made a stupid mistake here.

I have a NavigationAgent2D on which I am setting a radius. However, that radius does not seem to be affecting the paths provided by $NavigationAgent.get_next_location() and $NavigationAgent.get_nav_path(). The returned paths are hugging tight to corners and walls, whereas I would expect them to be offset by the radius value.

![image](https://user-images.githubusercontent.com/2744848/165428828-299a99bd-08ab-4a84-ad1b-1cf70555ce48.png)

### Steps to reproduce

Download and run the reproduction project. Despite a radius being set in the navigationagent (as confirmed with a print output), it hugs right to corners as if radius was 0.

### Minimal reproduction project

[Learning 3.zip](https://github.com/godotengine/godot/files/8568309/Learning.3.zip)

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/60354 (possible duplicate)?

**Comment by nathanielhudson** [comment]
Maybe? #60354 seems to show a 3D navmesh correctly baking the safety radius, but then the NavigationAgent3D is traveling outside the navmesh for some reason. #57486 seems to show a NavigationAgent2D responding in apparently suboptimal ways to a NavigationObstacle2D - although it is correctly using the radius to avoid the NavigationObstacle2D.

This issue is a NavigationAgent2D not employing the radius when navigating around a NavigationPolygonInstance. I'm not using any NavigationObstacle2D.

(Not saying they're not related under the hood, just wanted to document what I perceived the differences between the issues to be)

**Comment by Scony** [comment]
@Calinou I think the milestone for this one should be `3.5`.

Regarding the issue itself, it's unrelated to `3D`.

The issue here is due to the fundamental problem of `2D` navigation in Godot which is: **there is no baking stage**.

In `3D` for example, we have to bake navigation mesh before we can use it. One of the parameters we provide to the baking stage is `agent_radius` which actually should be called `max_agent_radius`. It tells the navmesh calculator what is the largest possible agent radius which can be used in the context of this particular navmesh. The navmesh calculator uses it to shrink the navmesh in a way that the largest agent will fit into the original geometry. So e.g. if we have a `10x10` plane and `max_agent_radius=2` the resulting navmesh will be shrunk by 2 units in every direction, so it will be `6x6` so that every agent with `R<=2` can safely walk `10x10` plane by moving its center within that `6x6` area.

In `2D` navigation we have no baking stage and thus the user must take care of polygons himself. This is fine if one is creating navigation polygons from code, but the problem strikes if one uses a tilemap. In that case, one would have to introduce extra tiles with navigation margins which is so cumbersome that I consider it virtually impossible.

The only way to overcome this problem would be IMO to introduce a new option to the `NavigationAgent2D` so that it uses non-optimized paths for movement. In the case of tilemap the paths would go through tile centers like this (red line):
![xxxx](https://user-images.githubusercontent.com/1207385/165987991-d80155c2-4d10-4571-9cf4-71f9aa864501.png)

@nathanielhudson would such an option be satisfactory to you?

**Comment by Calinou** [comment]
> I think the milestone for this one should be 3.5.

This sounds like a large thing to tackle for 3.5, so I'd prefer keeping the milestone to `3.x` to not disappoint people in case this needs to be pushed back.

**Comment by Scony** [comment]
> > I think the milestone for this one should be 3.5.
> 
> This sounds like a difficult problem to tackle for 3.5, so I'd prefer keeping the milestone to `3.x` to not disappoint people in case this needs to be pushed back.

ok, sounds good

**Comment by nathanielhudson** [comment]
@Scony - Tile centers would be good for my use case, but it might be a problem for others - tile centers assumes that the entities are approximately the same size as the agents - If I had small tiles and big agents, it may not work.

I feel that ideally we'd have either the navmesh or agent be smart enough to take the agent's radius into account. Spitballing here... it would be neat if the navmesh or agent could internally cache a modified version of the navmesh geometry contracted by the agents' sizes. For example, my 10px agent goes to make a path and creates a new version of the navmesh geometry offset_polygon-ed by 10px. If that agent asks for another path the cached geometry could be re-used. If a 20px agent asked for a path, a new version with a 20px offset would be created and cached. Don't know if that's too much complexity or would have weird performance implications though.

**Comment by Scony** [comment]
> I feel that ideally we'd have either the navmesh or agent be smart enough to take the agent's radius into account. Spitballing here... it would be neat if the navmesh or agent could internally cache a modified version of the navmesh geometry contracted by the agents' sizes. For example, my 10px agent goes to make a path and creates a new version of the navmesh geometry `offset_polygon`-ed by 10px. If that agent asks for another path the cached geometry could be re-used. If a 20px agent asked for a path, a new version with a 20px offset would be created and cached. Don't know if that's too much complexity or would have weird performance implications though.

IMO that's only doable with **baking stage** implemented.

**Comment by nathanielhudson** [comment]
Fair. At risk of asking a dumb question, is there a reason why 2D has no bake?

**Comment by Scony** [comment]
> Fair. At risk of asking a dumb question, is there a reason why 2D has no bake?

It was never implemented. Till `3.5` Godot navigation system was very straightforward.

**Comment by viksl** [comment]
@smix8 
Is there a way to use this workaround with a tilemap too, please?

**Comment by golddotasksquestions** [comment]
This is very disappointing. I have been looking avoiding the current Godot3 Navigation2D exactly because entities get stuck on corners and was hoping the new navigation solution in Godot 3.5 and Godot4 to solve this long standing issue. It's great we now have it built-into 3D navigation, but the vast majority of Godot users are still using Godot for 2D as recent poll shows.

**Comment by AlejandroMonteseirin** [comment]
I have the same problem in Godot 3.5, Tile map navigation have problems because characters get stuck in corners or try to pass in 3px routes when they have a size of 100px.Also, navigation agent avoidance radius only works with obstacles, not with the tilemap navigation. I workaround the navigation by removing a chunk of navigation around the walls, but only will work if all agents have the same size and have a lot of problems. Any solution?
![image](https://user-images.githubusercontent.com/33956661/183266628-74648837-aaf9-4828-9d80-64b815ecaa8c.png)

**Comment by smix8** [comment]
@AlejandroMonteseirin 
The only solution is to fix your navpolygon by adding enough margin for whatever actor size should use the navigationmesh for pathfinding. The navigationmesh describes the safe-zone for the actors center position and is the baked result with all the information about collision / scene layout / actor size and not the other way around.

**Comment by timothyqiu** [comment]
As a workaround, you can set the navigation region to be half the size of the tile:

![ksnip_20220909-191555](https://user-images.githubusercontent.com/372476/189338113-a3d42638-c615-4046-affd-7c9520ce677a.png)

Then set "Edge Connection Margin" to half the tile size. The project settings `navigation/2d/default_edge_connection_margin`, and also available on the Navigation node.

![Peek 2022-09-09 19-14](https://user-images.githubusercontent.com/372476/189337902-649f04ec-f23a-4da4-b85d-73ee25371790.gif)

**Comment by stats** [comment]
That is a nice workaround. If you have all entities of the same size it may work great. 

If you have entities of different sizes you may still have some issues. IMO NavigationAgent2D will still need to have a radius setting so that different sized agents can use the same navmesh.

**Comment by ka0s420** [comment]
Unfortunately that work around doesn't work for my situation, as the walls in my map are much thinner than the total tile size, so if i make a margin big enough to connect the little island navmesh pieces like you have in the gif above, then the agents can form paths through the walls. I'm also quite nonplussed at how you're able to get such uniform perfect navmeshes on ur tiles, I couldn't find a grid feature when trying to make mine. there is a snap feature, but no grid:

![tileNavProblem](https://user-images.githubusercontent.com/23391250/197050322-aab13302-2c5c-48f0-81d8-a1d6eee69c63.PNG)

**Comment by smix8** [comment]
Working on 2D NavigationMesh Baking, progress see video below.

[![2D NavigationMesh Baking](https://user-images.githubusercontent.com/52464204/216736340-59e098cf-09f2-43ff-876b-abef4d435ae4.jpg)](https://www.youtube.com/watch?v=w6ejSR830XY)

**Comment by smix8** [comment]
Added parser to bake a NavigationRegion2D NavigationPolygon from TileMap / TileSet

Works the same as NavigationRegion3D works with GridMap.

![tilemap_2d_baking_01](https://user-images.githubusercontent.com/52464204/217048075-9cd9c4c5-5dfa-4c76-bf86-ea64331caeca.gif)

**Comment by hornetDC** [comment]
@smix8 any updates on this?

**Comment by smix8** [comment]
@hornetDC 
Will be part of pr https://github.com/godotengine/godot/pull/70724 if you want to track it. It is not part of the pr yet as I did not have time to push the update from my local working branch yet.

**Comment by smix8** [comment]
@smks Godot 4.1+ only.

**Comment by Vivalzar** [comment]
1. This issue is marked as closed. Why? Is there new APIs? Because it seems it is still present in Godot 4.2.

2. And so, I tried to implement the workaround given here by @smix8. I used the example from the [2D Navigation Overview](https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_introduction_2d.html#setup-for-2d-scene). I can't seem to make it work. The radius stays at 0. What am I missing?

Here is the code I'm using:

```gdscript
extends CharacterBody2D

var movement_speed: float = 20.0
var movement_target_position: Vector2 = Vector2(350.0, 190.0)

@onready var navigation_agent: NavigationAgent2D = $NavigationAgent2D
@onready var navigation_region: NavigationRegion2D = $"../NavigationRegion2D"

func _ready():
	# These values need to be adjusted for the actor's speed
	# and the navigation layout.
	navigation_agent.path_desired_distance = 4.0
	navigation_agent.target_desired_distance = 4.0

	# Make sure to not await during _ready.
	call_deferred("actor_setup")

func actor_setup():
	### MY ADDITION BELOW
	var navigation_map: RID = NavigationServer3D.map_create()
	var navigation_mesh: NavigationMesh = navigation_region.navigation_polygon.get_navigation_mesh()
	navigation_mesh.agent_radius = 10.0
	NavigationMeshGenerator.bake(navigation_mesh, self)
	NavigationServer3D.region_set_navigation_mesh(navigation_region.get_region_rid(), navigation_mesh)
	NavigationServer3D.region_set_map(navigation_region.get_region_rid(), navigation_map)
	NavigationServer3D.map_set_cell_size(navigation_map, 1.0)
	NavigationServer3D.map_force_update(navigation_map)
	navigation_agent.set_navigation_map(navigation_map)
	### MY ADDITION ABOVE

	# Wait for the first physics frame so the NavigationServer can sync.
	await get_tree().physics_frame

	# Now that the navigation map is no longer empty, set the movement target.
	set_movement_target(movement_target_position)

func set_movement_target(movement_target: Vector2):
	navigation_agent.target_position = movement_target

func _physics_process(_delta):
	if navigation_agent.is_navigation_finished():
		return

	var current_agent_position: Vector2 = global_position
	var next_path_position: Vector2 = navigation_agent.get_next_path_position()

	var new_velocity: Vector2 = next_path_position - current_agent_position
	new_velocity = new_velocity.normalized()
	new_velocity = new_velocity * movement_speed

	velocity = new_velocity
	move_and_slide()
```

**Comment by Vivalzar** [comment]
I could make it work eventually. Here is the solution I used:

```gdscript
func actor_setup():
	### MY ADDITION BELOW
	var radius: float = -10.0
	var tab: Array[PackedVector2Array]
	var polygon = NavigationPolygon.new()
	var outline = navigation_region.navigation_polygon.get_outline(0)
	tab = Geometry2D.offset_polygon(outline, radius)
	polygon.add_outline(tab[0])
	polygon.make_polygons_from_outlines()
	navigation_region.navigation_polygon = polygon
	### MY ADDITION ABOVE

	# Wait for the first physics frame so the NavigationServer can sync.
	await get_tree().physics_frame

	# Now that the navigation map is no longer empty, set the movement target.
	set_movement_target(movement_target_position)
```
